src/Simulation/CsharpGeneration/Context.fs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Microsoft.Quantum.QsCompiler.CsharpGeneration
+
+open System
+open System.Collections.Generic
+open System.Collections.Immutable
+open System.Linq
+
+open Microsoft.Quantum.QsCompiler
+open Microsoft.Quantum.QsCompiler.DataTypes
+open Microsoft.Quantum.QsCompiler.ReservedKeywords
+open Microsoft.Quantum.QsCompiler.SyntaxTree
+open Microsoft.Quantum.QsCompiler.Transformations
+
+
+type internal DeclarationPositions() = 
+    inherit SyntaxTreeTransformation<NoScopeTransformations>(new NoScopeTransformations())
+
+    let mutable currentSource = null
+    let declarationLocations = new List<NonNullable<string> * (int * int)>()
+
+    member this.DeclarationLocations = 
+        declarationLocations.ToLookup (fst, snd)
+
+    static member Apply (syntaxTree : IEnumerable<QsNamespace>) = 
+        let walker = new DeclarationPositions()
+        for ns in syntaxTree do 
+            walker.Transform ns |> ignore
+        walker.DeclarationLocations
+
+    override this.onSourceFile file = 
+        currentSource <- file.Value
+        file
+
+    override this.onLocation (loc : QsLocation) = 
+        if currentSource <> null then
+            declarationLocations.Add (NonNullable<string>.New currentSource, loc.Offset)
+        loc
+
+
+type CodegenContext = { 
+    assemblyConstants       : IDictionary<string, string>
+    allQsElements           : IEnumerable<QsNamespace>
+    allUdts                 : ImmutableDictionary<QsQualifiedName,QsCustomType>
+    allCallables            : ImmutableDictionary<QsQualifiedName,QsCallable>
+    declarationPositions    : ImmutableDictionary<NonNullable<string>, ImmutableSortedSet<int * int>>
+    byName                  : ImmutableDictionary<NonNullable<string>,(NonNullable<string>*QsCallable) list>
+    current                 : QsQualifiedName option
+    signature               : ResolvedSignature option
+    fileName                : string option
+    unitTests               : ILookup<NonNullable<string>, string> // for each namespace contains the targets on which unit tests need to be executed
+} 
+    with
+    static member internal Create (syntaxTree, assemblyConstants, mayContainTests) =        
```

Comment:
I dont' think this should be a parameter. If tests are defined, it should just generate the code for them.